core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -116,25 +117,65 @@ class GroupCoordinator(val brokerId: Int,
     if (sessionTimeoutMs < groupConfig.groupMinSessionTimeoutMs ||
       sessionTimeoutMs > groupConfig.groupMaxSessionTimeoutMs) {
       responseCallback(joinError(memberId, Errors.INVALID_SESSION_TIMEOUT))
-    } else {
-      // only try to create the group if the group is not unknown AND
-      // the member id is UNKNOWN, if member is specified but group does not
-      // exist we should reject the request
-      groupManager.getGroup(groupId) match {
-        case None =>
-          if (memberId != JoinGroupRequest.UNKNOWN_MEMBER_ID) {
-            responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
-          } else {
-            val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
-            doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-          }
+      return
+    }
 
-        case Some(group) =>
+    groupManager.getGroup(groupId) match {
+      case None =>
+        // only try to create the group if the group is UNKNOWN AND
+        // the member id is UNKNOWN, if member is specified but group does not
+        // exist we should reject the request.
+        if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
+          doUnknownJoinGroup(group, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+        } else {
+          responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
+        }
+
+      case Some(group) =>
+        if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          doUnknownJoinGroup(group, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+        } else {
           doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-      }
+        }
     }
   }
 
+  private def doUnknownJoinGroup(group: GroupMetadata,
+                                 requireKnownMemberId: Boolean,
+                                 clientId: String,
+                                 clientHost: String,
+                                 rebalanceTimeoutMs: Int,
+                                 sessionTimeoutMs: Int,
+                                 protocolType: String,
+                                 protocols: List[(String, Array[Byte])],
+                                 responseCallback: JoinCallback): Unit = {
+    group.inLock {
+      if (group.is(Dead)) {
+        // if the group is marked as dead, it means some other thread has just removed the group
+        // from the coordinator metadata; this is likely that the group has migrated to some other
```

Comment:
nit: `this is likely` -> `it is likely`?

Reply:
fixed all the comment occurrence.